feat: @W-18480458: Implementation of update references across Omnistudio components

sf-kishore-kurri opened the pull request:
Changes:
- Handling references updates post migration clean up changes
- Added logic to perform migrate in correct order DM -> IP -> OS and FC
- Created the namingRegistry to store old  and cleanup name for lookup
- Added the test cases

### What does this PR do?

### What issues does this PR fix or reference?

sf-kishore-kurri committed:
feat: @W-18480458: Implementation of update references across Omnistudio components

Changes:
- Handling references udpates post migration clean up changes
- Added logic to perform migrate in correct order DM -> IP -> OS and FC
- Created the namingRegistry to store old  and cleanup name for lookup
- Added the test cases

sf-kishore-kurri committed:
feat: @W-18480458: Implementation of update references across Omnistudio components

Changes:
- Handling references udpates post migration clean up changes
- Added logic to perform migrate in correct order DM -> IP -> OS and FC
- Created the namingRegistry to store old  and cleanup name for lookup
- Added the test cases

sf-aastha-paruthi commented:
In case of OS and FC, both can have dependencies on each other. Is the correct order change mentioned taking care of that.

sf-kishore-kurri commented:
> e correct order change mentioned taking care

The order of migration happens like DM, IP, OS and FC so if OS has any FC dependency it will be taken care by the namingRegistry because it has the oldName <-> cleanedName so it will update automatically.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -440,6 +494,114 @@ export default class Migrate extends OmniStudioBaseCommand {
+  /**
+   * Query DataMappers from the org
+   */
+  private async queryDataMappers(conn: any, namespace: string): Promise<any[]> {
```
The functions to query and fetch the datamappers and other components are already written in each class like getAllDatamappers.
Please reuse those. A bit refactoring might be needed.

sf-kishore-kurri replied:
The getAllDatamappers created as private one only accessible within each migration component type and i'm preparing registry before migration so created these query utility APIs for accessing the pre-migrated data

snehaljha-sf replied:
We can change teh declaration visibility or add new method. It's better to have all the queries to be defined at same/similar place

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -440,6 +494,114 @@ export default class Migrate extends OmniStudioBaseCommand {
+  /**
+   * Query OmniScripts/Integration Procedures from the org
+   */
+  private async queryOmniScripts(conn: any, namespace: string, isProcedure: boolean): Promise<any[]> {
```
Please reuse the already existing functions.

sf-kishore-kurri replied:
The component mapping query only fetching 2-3 columns where as migration need more columns for data copy so we can leave the mapping query separate from data migration ones  @snehaljha-sf

sf-aastha-paruthi commented on src/migration/NameMappingRegistry.ts:
```diff
@@ -0,0 +1,401 @@
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public preProcessComponents(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Please try avoiding es lint checks

sf-kishore-kurri replied:
will check

sf-aastha-paruthi commented on src/migration/flexcard.ts:
```diff
@@ -773,81 +775,261 @@ export class CardMigrationTool extends BaseMigrationTool implements MigrationToo
+    if (type === Constants.DataRaptorComponentName || type === 'DataRaptor') {
+      // Handle DataRaptor using registry
+      const originalBundle = dataSource.value?.bundle || '';
+      if (originalBundle && this.nameRegistry.hasDataMapperMapping(originalBundle)) {
```
In what scenarios will the nameRegistry might not have the nameMapping.

sf-kishore-kurri replied:
In most cases, registry should have the key->value but added fallback so that any missing case also update will be done based on cleaned Name

sf-aastha-paruthi commented on src/migration/flexcard.ts:
```diff
@@ -773,81 +775,261 @@ export class CardMigrationTool extends BaseMigrationTool implements MigrationToo
+      if (originalBundle && this.nameRegistry.hasDataMapperMapping(originalBundle)) {
+        dataSource.value.bundle = this.nameRegistry.getDataMapperCleanedName(originalBundle);
+      } else {
+        dataSource.value.bundle = this.cleanName(originalBundle);
```
Does the nameregistry need to be updated?

sf-kishore-kurri replied:
Ideally we shouldn't see any misses because registry is prepared based on pre-migrated data and else logic is fallback to avoid not clean up names

snehaljha-sf replied:
We should add some logs to let us know about any such miss

sf-kishore-kurri replied:
sure @snehaljha-sf

sf-aastha-paruthi commented on src/migration/NameMappingRegistry.ts:
```diff
@@ -0,0 +1,401 @@
+ * Centralized registry for tracking name mappings and managing dependency updates
+ * during Omnistudio component migration.
+ */
+export class NameMappingRegistry {
```
What is the use case of this NameMappingRegistry.

sf-kishore-kurri replied:
NamingRegistry is in memory storage which is prepared before migration for all component types to maintain key, value between oldName -> cleanName which can be used for update references across different migration component types

sf-aastha-paruthi commented on src/migration/NameMappingRegistry.ts:
```diff
@@ -0,0 +1,401 @@
+    this.nameMappings.set(key, mapping);
+
+    // Store in type-specific maps for quick lookup
+    switch (mapping.componentType) {
```
The same can be derived at runtime. Why do we need to store the same.

sf-aastha-paruthi commented:
> > e correct order change mentioned taking care
> 
> The order of migration happens like DM, IP, OS and FC so if OS has any FC dependency it will be taken care by the namingRegistry because it has the oldName <-> cleanedName so it will update automatically.

This will not ensure the correct reference update. The unmigrated or Error items will not be handled correctly.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -128,58 +129,44 @@ export default class Migrate extends OmniStudioBaseCommand {
-      (result) => result.name !== Constants.GlobalAutoNumberComponentName
-    );
+    // We need to truncate the standard objects first (in reverse order for cleanup)
+    let objectMigrationResults = await this.truncateObjects([...migrationObjects].reverse(), debugTimer);
```
Please update the name. This does not seem to be migration results.

sf-kishore-kurri replied:
its existing code and truncate is returning MigratedObject[]

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -374,29 +408,41 @@ export default class Migrate extends OmniStudioBaseCommand {
+  /**
+   * Get migration objects in the correct dependency order:
+   * 1. DataMappers (lowest dependencies)
+   * 2. OmniScripts/Integration Procedures
```
Update the comment to reflect the correct order.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -420,18 +467,25 @@ export default class Migrate extends OmniStudioBaseCommand {
         case Constants.GlobalAutoNumber:
-          migrationObjects.push(new GlobalAutoNumberMigrationTool(namespace, conn, this.logger, messages, this.ux));
+          migrationObjects.push(
+            new GlobalAutoNumberMigrationTool(namespace, conn, this.logger, messages, this.ux, nameRegistry)
```
If this is independent item, do we still have dependency on nameRegistry.

sf-kishore-kurri replied:
For now, I'm passing the namingRegistry to GlobalAutoNumber migration process as well so that we can use it in future for any reference updates

sf-kishore-kurri commented:
> e correct reference update. T

The failure/ error ones should be remigrated then name registry will take care of it of update references

snehaljha-sf commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -129,59 +130,44 @@ export default class Migrate extends OmniStudioBaseCommand {
-    // Register the migration objects
+    // Initialize the name mapping registry and pre-process all components
+    const nameRegistry = NameMappingRegistry.getInstance();
+    nameRegistry.clear(); // Clear any previous mappings
```
We don't need to clear the mappigns as we are not persisting it in any file based storage

sf-kishore-kurri replied:
clearing post init

snehaljha-sf commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -129,59 +130,44 @@ export default class Migrate extends OmniStudioBaseCommand {
+    nameRegistry.clear(); // Clear any previous mappings
+
+    Logger.log(messages.getMessage('startingComponentPreProcessing'));
+    await this.preProcessAllComponents(nameRegistry, namespace, conn, migrateOnly);
```
nameRegistry follows a singleton pattern, we don't need to add this dependency while calling a function

sf-kishore-kurri replied:
you mean NameMappingRegistry.getInstance() should be good instead of passing

snehaljha-sf commented on src/migration/NameMappingRegistry.ts:
```diff
@@ -0,0 +1,401 @@
+  private dependencyReferences: DependencyReference[] = [];
+
+  // Type-specific mappings for quick lookup
+  private dataMapperMappings: Map<string, string> = new Map(); // original -> cleaned
```
why do we need componentWise name mapping if we already have common name mapping with prefix

sf-kishore-kurri replied:
@snehaljha-sf component mapping is for faster lookup

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -240,6 +226,51 @@ export default class Migrate extends OmniStudioBaseCommand {
+        // Use ProjectPathUtil for APEX project folder selection (matches assess.ts logic)
+        projectPath = await ProjectPathUtil.getProjectPath(messages, true);
+        targetApexNamespace = await this.getTargetApexNamespace(objectsToProcess, targetApexNamespace);
+        await this.handleExperienceSitePrerequisites(
```
This has been moved to premigrate class.

sf-kishore-kurri replied:
yes, its merge conflict issue taking care of it

sf-kishore-kurri committed:
fix: @W-18480458: Code review comments and fixes

sf-kishore-kurri committed:
fix: @W-18480458: Build failures

sf-kishore-kurri committed:
fix: @W-18480458: Additional fixes

Changes:
- Handling of skipping Angular Omniscript dependencies FC and OS
- Split the Omniscript and IP migration and also corrected the labels
- Added labels and test cases

sf-kishore-kurri committed:
fix: @W-18480458: Assessment issues

sf-kishore-kurri committed:
fix: @W-18480458 : Fix for label correction

sf-kishore-kurri committed:
fix: @W-18480458: OMA assess split for OS and IP

sf-kishore-kurri commented:
@snehaljha-sf @sf-aastha-paruthi Please review the changes

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -129,59 +130,43 @@ export default class Migrate extends OmniStudioBaseCommand {
+    let objectMigrationResults = await this.truncateObjects([...migrationObjects].reverse(), debugTimer);
     const allTruncateComplete = objectMigrationResults.length === 0;
 
     if (allTruncateComplete) {
```
Please verify by running migration and assessment with only os, fc.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -240,6 +225,59 @@ export default class Migrate extends OmniStudioBaseCommand {
+          isExperienceBundleMetadataAPIProgramaticallyEnabled
+        );
+        Logger.logVerbose(
+          'The objects to process after handleExpSitePrerequisite are ' + JSON.stringify(objectsToProcess)
```
This has been moved to message.json

sf-kishore-kurri replied:
Even on latest, we aren't using the messages

sf-aastha-paruthi commented on src/commands/omnistudio/migration/assess.ts:
```diff
@@ -208,7 +208,8 @@ export default class Assess extends OmniStudioBaseCommand {
       await this.assessDataRaptors(assesmentInfo, namespace, conn);
       await this.assessFlexCards(assesmentInfo, namespace, conn, allVersions);
-      await this.assessOmniScripts(assesmentInfo, namespace, conn, allVersions, OmniScriptExportType.All);
+      await this.assessOmniScripts(assesmentInfo, namespace, conn, allVersions, OmniScriptExportType.OS);
```
Shouldn't the order of assessment and migration be same, given that there can be FC->OS and OS->FC dependencies.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -340,18 +380,37 @@ export default class Migrate extends OmniStudioBaseCommand {
     allVersions: any
   ): MigrationTool[] {
     if (!migrateOnly) {
+      // Correct order: DataMapper -> OmniScript/IP -> FlexCard -> GlobalAutoNumber
```
nit, remove duplicate comment, comment in line 383 already covers the ordering.

sf-aastha-paruthi commented on src/commands/omnistudio/migration/migrate.ts:
```diff
@@ -406,6 +467,137 @@ export default class Migrate extends OmniStudioBaseCommand {
+  /**
+   * Query FlexCards from the org
+   */
+  private async queryFlexCards(conn: any, namespace: string): Promise<any[]> {
```
All the 4 functions are making the same call to query and then return the resuts, just the query is different. The query can be separated and the other part can be put to a common function to avoid duplicacy.

sf-aastha-paruthi commented on src/migration/omniscript.ts:
```diff
@@ -196,10 +200,11 @@ export class OmniScriptMigrationTool extends BaseMigrationTool implements Migrat
     try {
-      Logger.log(this.messages.getMessage('startingOmniScriptAssessment'));
+      const exportComponentType =
+        this.exportType === OmniScriptExportType.IP ? 'Integration Procedures' : 'Omniscripts';
```
getName() can be used.

sf-aastha-paruthi commented on src/migration/omniscript.ts:
```diff
@@ -231,7 +236,9 @@ export class OmniScriptMigrationTool extends BaseMigrationTool implements Migrat
 
-    const progressBar = createProgressBar('Assessing', 'Omniscript and Integration Procedure');
+    const progressBarType: ComponentType =
+      this.exportType === OmniScriptExportType.IP ? 'Integration Procedures' : 'Omniscripts';
```
getName()

sf-kishore-kurri replied:
Done

sf-kishore-kurri committed:
fix: @W-18480458 : Resolved the conflicts and code review fixes

sf-kishore-kurri committed:
fix: @W-18480458: more fixes

sf-kishore-kurri commented:
@snehaljha-sf @sf-aastha-paruthi addressed the review comments
